pgjdbc/src/main/java/org/postgresql/copy/PGCopyInputStream.java

```diff
@@ -85,27 +88,32 @@ public int read(byte[] buf) throws IOException {
 
   public int read(byte[] buf, int off, int siz) throws IOException {
     checkClosed();
-    int got = 0;
+    int bytesRead = 0;
     boolean didReadSomething = false;
-    while (got < siz && (didReadSomething = gotBuf())) {
-      buf[off + got++] = this.buf[at++];
+    while (bytesRead < siz && (didReadSomething = isBufAvailable())) {
+      buf[off + bytesRead++] = this.buf[curPosition++];
     }
-    return got == 0 && !didReadSomething ? -1 : got;
+    // need to check both. In the loop above on the first iteration
+    // bytesRead==0 and didReadSomething=false indicates -1
+    return bytesRead == 0 && !didReadSomething ? -1 : bytesRead;
   }
 
   public byte[] readFromCopy() throws SQLException {
     byte[] result = buf;
```

Comment:
Why is this being assigned here? It is always assigned to something else below.

Reply:
ah - not all cases below.
If return from ‘isBufAvailable’ is true and after that call curPosition is not 0 or ‘len < buf.length’, then this current buf value is used.
That seems like it could be incorrect. The call to isBufAvailable could result in changing to a different buf. Specifically if currently exactly at end of buf.

Reply:
I think this is the bug. Even my comments make note of isBufAvailable doing too much and changing buf seems like a bad thing here.

Reply:
I think we want something like this:

```
    final byte[] result;
    try {
      if (isBufAvailable()) {
        // if the entirety of current buffer is available, just hand back reference to it
        if (curPosition == 0 && len = buf.length) {
          result = buf;
        } else {
          result = Arrays.copyOfRange(buf, curPosition, curPosition + len);
        }
        //either way, we have completely used up the buffer
        curPosition = len; 
      } else {
        result = null;
      }
    } catch (IOException ioe) {
      throw new PSQLException(GT.tr("Read from copy failed."), PSQLState.CONNECTION_FAILURE);
    }
    return result;
```

Reply:
ugh - should be:

`result = Arrays.copyOfRange(buf, curPosition, len);`

perhaps consider renaming `len` to be `toPosition` or something similar.